### PR TITLE
Restructure, Add Subject Cache, Context Calls, TLS Config & More

### PR DESCRIPTION
**Pull request by AtakanColak**
Hiya, I've been working on a restructuring of this library, and wanted to create a draft for people who may work on the same thing and maybe put some ideas.

TODO List 
- [x] Fix too many open files error of HTTP requests with a semaphore
- [x] Fix race condition of CachingEnabled and CodecCreationEnabled
- [x] Get Schemas by Subject, in a cached manner
- [ ] Do HTTP calls with contexts
- [ ] Expose TLS Config as an Option
- [ ] Tidy up, submit with least amount of breaking changes

### CODE REVIEW

**Comment by AtakanColak** [comment]
@riferrei I will work on seperate issues and commits for the master repository that don't have breaking changes. I started with #33. However, some of the components I worked on do include breaking changes, so I'm not sure what to do for that case without disrupting people who already use this library. I think we have three options here;

1. My breaking changes are kept in my fork and I just work on my fork for my own needs. 
2. We plan for a v2 folder together in this repository with predetermined breaking changes and I restructure my commits that way. 
3. We plan together and publish breaking changes to master.

What do you think we should do?

**Comment by riferrei** [comment]
Hi @AtakanColak, sorry for taking too much time to respond. First of all, thanks for your contributions. They are great. Regarding how do we handle breaking changes; I would favor option number 3 but start creating releases for this projects. This way people could make sure which specific release they are using while we advance the code with new features and bug fixes.

I just made you contributor. So you can control the lifecycle of your PRs =)

**Comment by AtakanColak** [comment]
@riferrei thank you, I'm not comfortable with managing PRs all by myself so I'll ask for your and others' review & test before merging them. I'm not really sure on how to keep older versions by tagging releases, and being able to apply bugfixes on them. I will try to do research on how to do this.

File: mockSchemaRegistryClient_test.go [hunk 1]
```diff
@@ -2,82 +2,80 @@ package srclient
...
-"\"doc\": \"Sample schema to help you get started.\", " +
-"\"fields\": [    " +
-"{   " +
-"\"name\": \"aField\"," +
-"\"type\": \"int\", " +
-"\"doc\": \"The int type is a 32-bit signed integer.\"   " +
-"}" +
-"]" +
-"}"
+	"\"type\": \"record\",  " +
```
Comment (tomarrell) [inline mockSchemaRegistryClient_test.go:17]: Probably more preferable to use a string literal for defining these. https://golang.org/ref/spec#String_literals

Comment (AtakanColak) [reply mockSchemaRegistryClient_test.go:17]: Thank you for your reply, I'm re-adapting this work at [v1.0.0 branch](https://github.com/AtakanColak/srclient/tree/v1.0.0-alpha) though it is not near completion yet.

**Comment by AtakanColak** [comment]
Closed in favour of #49

**Pull request closed without merging**